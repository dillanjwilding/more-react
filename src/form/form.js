import React, { Children, cloneElement, useEffect, useState } from 'react'
import PropTypes from 'prop-types'

// @todo: Track down extra renders and remove them
// @todo: Add layout and formatting
// @todo: Add loading from JSON
// @todo: Add validation
// @todo: I shouldn't need to rely on displayName. There should be another way to reflexively get a component's name
// @todo: Export default Form
export const Form = ({ children, onChange, onSubmit, preventDefault = true, ...props }) => {
  // const [loading, setLoading] = useState(true)
  const [step, setStep] = useState(0)
  const [values, setValues] = useState({})
  useEffect(() => {
    // setLoading(false)
  }, [])
  const handleSubmit = event => {
    if (preventDefault) event.preventDefault()
    if (onSubmit) onSubmit(values)
    console.log('submitted')
  }
  /* Children.map(children, (child, index) => {
    console.log(child.name)
  }) */
  const steps = Children.toArray(children).filter(child => child.type.displayName === 'Step').length
  return (
    <form {...props} onSubmit={handleSubmit}>
      {Children.map(children, (child, index) => {
        let childProps = {
          key: index, // not the best practice
          onChange: event => {
            event.persist()
            if (onChange) onChange()
            setValues(values => ({ ...values, [event.target.name]: event.target.value }))
            console.log(values)
          }
        }
        if (child.type.displayName === 'Step') {
          childProps = {
            ...childProps,
            hidden: index !== step,
            step,
            steps,
            setStep
          }
        }
        return cloneElement(child, childProps)
        /* let type = child.props.type
          if (child.props.type && typeof child.props.type !== 'string') {
            child.props.type.name.toLowerCase()
          }
          const id = child.props.children.split(' ').join('_').toLowerCase()
          if (['select', 'textarea'].includes(type)) {
            return <type id={id} name={id} />
          } else {
            return <input type={type} id={id} name={id} />
          } */
      })}
    </form>)
}
Form.displayName = 'Form' // @todo: Clean this up
Form.propTypes = {
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  preventDefault: PropTypes.bool
}
