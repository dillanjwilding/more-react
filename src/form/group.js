import React from 'react'

export const FormGroup = ({ children, ...props }) => {
  // @todo: Use fieldset and/or form-label-group for floating labels?
  return (
    <div className='form-group'>{children}</div>
  )
}
FormGroup.displayName = 'FormGroup' // @todo: Clean this up
