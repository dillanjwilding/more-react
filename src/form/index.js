import 'core-js'
import 'regenerator-runtime/runtime'

import 'bootstrap/dist/css/bootstrap.min.css'

// @todo: Track down extra renders and remove them
export * from './form.js'
export * from './step.js'
export * from './group.js'
export * from './fields.js'
