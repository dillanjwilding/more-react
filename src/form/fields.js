import PropTypes from 'prop-types'
import React from 'react'

import { FormGroup } from './group.js'

export const Field = ({ children, ...props }) => {
  if (!children) {
    throw Error('')
  }
  // const type = props.type.name.toLowerCase()
  const type = (props.type ? props.type.toLowerCase() : 'text')
  const id = children.split(' ').join('_').toLowerCase()
  return (
    <FormGroup>
      <label htmlFor={id}>{children}</label>
      <input type={type} id={id} name={id} className='form-control' placeholder={children} {...props} />
    </FormGroup>
  )
}
Field.displayName = 'Field' // @todo: Clean this up
export const Submit = ({ className, ...props }) => {
  return <input type='submit' className={className || 'btn btn-sm btn-success'} {...props} /> // @todo: Combine className?
}
Submit.displayName = 'Submit' // @todo: Clean this up
Submit.propTypes = {
  className: PropTypes.string
}
export const Email = props => {
  return <Field type='email' {...props} />
}
Email.displayName = 'Email' // @todo: Clean this up
export const Password = props => {
  return <Field type='password' {...props} />
}
Password.displayName = 'Password' // @todo: Clean this up
export const Area = ({ children, ...props }) => {
  // const type = props.type
  const id = children.split(' ').join('_').toLowerCase()
  return (
    <FormGroup>
      <textarea id={id} name={id} className='form-control' {...props} />
    </FormGroup>
  )
}
Area.displayName = 'Area' // @todo: Clean this up
export const Dropdown = ({ children, ...props }) => {
  // @todo: Figure out id/name
  return (
    <FormGroup>
      <select className='form-control' {...props}>
        <Option id='default' value=''>Select {props.name.split(' ').map(word => word.charAt(0).toUpperCase() + word.substring(1)).join(' ')}</Option>
        {children}
      </select>
    </FormGroup>
  )
}
Dropdown.displayName = 'Dropdown' // @todo: Clean this up
export const Option = ({ children, ...props }) => {
  const id = props.id || children.split(' ').join('_').toLowerCase()
  return <option {...props} id={id} name={id}>{children}</option> // @todo: Figure out id/name and/or value(?)
}
Option.displayName = 'Option' // @todo: Clean this up
