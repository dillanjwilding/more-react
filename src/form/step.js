import React, { Children, cloneElement } from 'react'

import { Submit } from './fields.js'

// @todo: Track down extra renders and remove them
export const Step = ({ hidden, children, onChange, step, steps, setStep, ...props }) => {
  const previous = e => {
    e.preventDefault()
    setStep(step - 1)
  }
  const next = e => {
    e.preventDefault()
    setStep(step + 1)
  }
  // Is hidden unnecessary because all steps != this step should be hidden.
  return (
    <div hidden={hidden}>
      {Children.map(children, (child, index) => {
        return cloneElement(child, {
          key: index, // not the best practice
          onChange
        })
      })}
      {(typeof step !== 'undefined' && step !== null) && (
        <>
          <PreviousButton previous={previous} step={step} />
          <NextButton next={next} step={step} steps={steps} />
          <SubmitButton step={step} steps={steps} />
        </>
      )}
    </div>)
}
Step.displayName = 'Step' // @todo: Clean this up

// @todo: Rewrite these
const PreviousButton = ({ previous, step }) => {
  if (typeof step !== 'undefined' && step !== null) {
    if (step !== 0) {
      return <button className='btn btn-sm btn-primary' onClick={previous}>Previous</button>
    }
  }
  return null
}

const NextButton = ({ next, step, steps }) => {
  if (typeof step !== 'undefined' && step !== null) {
    if (step < steps - 1) {
      return <button className='btn btn-sm btn-primary' onClick={next}>Next</button>
    }
  }
  return null
}

const SubmitButton = ({ step, steps }) => {
  if (typeof step !== 'undefined' && step !== null) {
    if (step === steps - 1) {
      return <Submit />
    }
  }
  return null
}

// const renderButtons = step => typeof step !== 'undefined' && step !== null

/* const NavigationButton = ({ action, Component, condition, step }) => {
  if (typeof step !== 'undefined' && step !== null) {
    if (condition) {
      return <Component onClick={action} />
    }
  }
} */
