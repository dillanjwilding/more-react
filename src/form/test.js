/* global describe, test, expect */
// @todo: Separate tests by component
import React from 'react'
import { shallow } from 'enzyme'
// import ReactTestRenderer from 'react-test-renderer'

import { Form } from './form.js'
import { Field } from './fields.js'
import { FormGroup } from './group.js'
import { Step } from './step.js'

describe('Test forms', () => {
  test('render correctly', () => {
    expect(shallow(<Form />)).toMatchSnapshot()
    // const renderer = ReactTestRenderer.create(<Form />)
    // expect(renderer.toJSON()).toMatchSnapshot()
  })
})

describe('Test form fields', () => {
  /* test('no parameters', () => {
    expect(shallow(<Field />)).toThrow(() => { throw new Error('') })
  }) */
  test('render correctly', () => {
    expect(shallow(<Field>Test</Field>)).toMatchSnapshot()
  })
})

describe('Test form groups', () => {
  test('render correctly', () => {
    expect(shallow(<FormGroup />)).toMatchSnapshot()
  })
})

describe('Test form steps', () => {
  test('render correctly', () => {
    expect(shallow(<Step />)).toMatchSnapshot()
  })
})
