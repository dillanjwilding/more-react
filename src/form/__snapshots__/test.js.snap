// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test form fields render correctly 1`] = `
<FormGroup>
  <label
    htmlFor="test"
  >
    Test
  </label>
  <input
    className="form-control"
    id="test"
    name="test"
    placeholder="Test"
    type="text"
  />
</FormGroup>
`;

exports[`Test form groups render correctly 1`] = `
<div
  className="form-group"
/>
`;

exports[`Test form steps render correctly 1`] = `<div />`;

exports[`Test forms render correctly 1`] = `
<form
  onSubmit={[Function]}
/>
`;
